src/charm.py

```diff
@@ -65,19 +73,67 @@ def _on_website_relation_joined(self, event):
                     'port': str(port)
                 })
 
+    def _on_metrics_endpoint_relation_created(self, event: RelationJoinedEvent):
+        username = metrics_username(event.relation)
+        password = secrets.token_urlsafe(16)
+        self.control_socket.add_metrics_user(username, password)
+
+        # Set up Prometheus scrape config
+        self.metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[{
+                # "job_name": "juju",
+                "metrics_path": "/introspection/metrics",
+                "scheme": "https",
+                "static_configs": [{"targets": ["*:17070"]}],
+                "basic_auth": {
+                    "username": f'user-{username}',
+                    "password": password,
+                },
+                "tls_config": {
+                    "ca_file": self.ca_cert(),
+                    "server_name": "juju-apiserver",
+                },
+            }],
+        )
+        self.metrics_endpoint.set_scrape_job_spec()
+
+    def _on_metrics_endpoint_relation_broken(self, event: RelationDepartedEvent):
+        # Remove metrics user
+        username = metrics_username(event.relation)
+        self.control_socket.remove_metrics_user(username)
 
-def api_port():
-    ''' api_port determines the port that the controller's API server is
-        listening on.  If the machine does not appear to be a juju
-        controller then None is returned.
+    def _agent_conf(self, key: str):
+        '''
+        _agent_conf reads a value from the agent.conf file on disk.
+        '''
```

Comment:
Style nit: we normally put one-line docstrings on a single line, like so (and unlike Go, Python style doesn't repeat the method name in the docstring):

```suggestion
        """Read a value (by key) from the agent.conf file on disk."""
```